### PR TITLE
EEP: System-wide Configuration by Embulk System Properties and Embulk Home

### PR DESCRIPTION
**Pull request by dmikurube**
This is an "retrospective" EEP (Embulk Enhancement Proposal) for the new Embulk System Properties and the Embulk Home directory.

### CODE REVIEW

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+Use-cases of System-wide Configurations
+----------------------------------------
+
+Let's see some actual use-cases of system-wide configurations.
+
+### Logging
+
+Logging-related configurations are a typical use-case of system-wide configurations. Embulk had configurations for the log level and the log file path in System Config.
+
+It, however, restricted logging in Embulk. System Config was available only after some initialization, then, logging was avaiable only after the initialization. Messages before the initialization had to be printed by `System.out` or `System.err` out of the control of the SLF4J logging framework.
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-embulk-system-properties-and-embulk-home.md:38]: avaiable -> available

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+Embulk Home
+------------
+
+The Embulk Home directory is determined at runtime, by the following rule, in priority from top to bottom.
+
+1. If an Embulk System Property `embulk_home` is set by the command-line option `-X`, it is the Embulk Home directory.
+    * It would be an absolute path, or a relative path from the current working directory.
+    * Ex. `java -jar embulk.jar -Xembulk_home=/var/tmp/foo run ...` (absolute path)
+    * Ex. `java -jar embulk.jar -Xembulk_home=.embulk2/bar run ...` (repative path)
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-embulk-system-properties-and-embulk-home.md:98]: repative  -> relative path

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+2. If an environment variable `EMBULK_HOME` is set, it is the Embulk Home directory.
+    * It must be an absolute path.
+    * Ex. `env EMBULK_HOME=/var/tmp/baz java -jar embulk.jar run ...` (absolute path)
+3. If neither 1 nor 2 matches, the Embulk Home directory is determined by searching from the current working directory as follows.
+    * The search starts from the current working directory, moving one by one toward the parent.
+    * In each directory, look for a directory, whose name is `.embulk/`, and which contains a regular file named `embulk.properties`.
+    * If such a directory `.embulk/` is found, it is the Embulk Home directory.
+    * If the current working directory is under the user's home directory (`user.home`), the search stops at the home directory.
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-embulk-system-properties-and-embulk-home.md:106]: What do you think about adding Java property explicitly? It's little confuse Embulk property or not.

**Review by hiroyuki-sato — COMMENTED** [review]
Basically LGTM. Could you check some comments?

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md (hunk 3 above)
Comment (dmikurube) [reply docs/eeps/draft-embulk-system-properties-and-embulk-home.md:106]: Let me confirm your intention. Do you mean :

1. Adding some explanation in this EEP about the relationship with "Java (System) Property"?
2. Adding another implemented rule to consider Java System Property (not just Embulk System Properties) to determine Embulk System Properties and Embulk Home?
3. Something else?

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md (hunk 1 above)
Comment (dmikurube) [reply docs/eeps/draft-embulk-system-properties-and-embulk-home.md]: Fixed it along with b66db914ce7facb66744d3167d76920db178f31b

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md (hunk 2 above)
Comment (dmikurube) [reply docs/eeps/draft-embulk-system-properties-and-embulk-home.md]: Fixed it along with b66db914ce7facb66744d3167d76920db178f31b

File: docs/eeps/draft-embulk-system-properties-and-embulk-home.md (hunk 3 above)
Comment (hiroyuki-sato) [reply docs/eeps/draft-embulk-system-properties-and-embulk-home.md]: `user.home` is used this place only.
I'm not familiar with Java, When I first read this EEP, I didn't realize `user.home` is the Java property, not Embulk System property. 

For example, Changing to "`user.home`(Java Property)", is helpful for me.

Comment (dmikurube) [reply docs/eeps/draft-embulk-system-properties-and-embulk-home.md]: Ah, got it! That's a reasonable point. Let me fix it.

Comment (dmikurube) [reply docs/eeps/draft-embulk-system-properties-and-embulk-home.md]: Fixed it in dedeb3c8ce8a9cd35f6ff5a62bc45eb3eba2aa94

**Review by hiroyuki-sato — APPROVED** [review]
LGTM👍

**Comment by dmikurube** [comment]
Thanks!

**Comment by dmikurube** [comment]
Let's merge it as EEP-4 as `Final`.

**Pull request merged**